### PR TITLE
[11.x] Add ability to override RouteServiceProvider

### PR DESCRIPTION
**Pull request by ahmedabdel3al**
I need to override `loadCacheRoutes` method inside `RouteServiceProvider` **as we did before laravel 11** but when i extend the framework RouteServiceProvider and add it inside the `bootstrap/providers.php` 
the framework RouteServiceProvider **always run last** 
i think we are forcing it here https://github.com/laravel/framework/pull/49732

### CODE REVIEW

**Comment by taylorotwell** [comment]
Can you describe instead what you are trying to accomplish as your end goal?

**Comment by ahmedabdel3al** [comment]
https://github.com/mcamara/laravel-localization 
We are using this package for many projects to translate routes. When we cache routes using their command, it generates many files in`bootstrap/cache` based on the locales supported by the application. 
For example, if we have 3 locales `[es, en, ar]`, this will generate 3 files:

- routes-v7_es.php
- routes-v7_en.php
- routes-v7_ar.php

Based on the current locale of the request, they load the correct cached routes file. So, we have overridden the RouteServiceProvider. 
https://github.com/mcamara/laravel-localization?tab=readme-ov-file#caching-routes
https://github.com/mcamara/laravel-localization/blob/master/src/Mcamara/LaravelLocalization/Traits/LoadsTranslatedCachedRoutes.php

File: src/Illuminate/Foundation/Configuration/ApplicationBuilder.php [hunk 1]
```diff
@@ -148,16 +148,17 @@ public function withRouting(?Closure $using = null,
...
         string $apiPrefix = 'api',
+        string $provider = AppRouteServiceProvider::class,
         ?callable $then = null)
     {
         if (is_null($using) && (is_string($web) || is_string($api) || is_string($pages) || is_string($health)) || is_callable($then)) {
             $using = $this->buildRoutingCallback($web, $api, $pages, $health, $apiPrefix, $then);
         }
 
-        AppRouteServiceProvider::loadRoutesUsing($using);
+        $provider::loadRoutesUsing($using);
```
Comment (rodrigopedra) [inline src/Illuminate/Foundation/Configuration/ApplicationBuilder.php:158]: As the custom provider should be a subclass of the framework's `RouteServiceProvider`, we can keep the static call to the base provider here.

IMO, the `RouteServiceProvider@loadRoutesUsing` static method should not be overridden

**Comment by rodrigopedra** [comment]
How about making `Application@resolveProvider` resolve the provider from the container? Instead of new'ing up the class directly from its class name?

https://github.com/laravel/framework/blob/a57dca1b039b02fca8a517f6df40f757009d2b86/src/Illuminate/Foundation/Application.php#L922-L925

This way, one could bind their custom provider on their `AppServiceProvider@register` method.

I would say, as customizing this appears to be very specific, would work better than adding one more parameter to the `withRouting()` method.

File: src/Illuminate/Foundation/Configuration/ApplicationBuilder.php (hunk 1 above)
Comment (ahmedabdel3al) [reply src/Illuminate/Foundation/Configuration/ApplicationBuilder.php:158]: Your Point Is valid, but i think both will do the same

**Comment by ahmedabdel3al** [comment]
@rodrigopedra I don't understand your point fully.

IMO,  when anyone open the `bootstrap/app.php` with this code 
```
Application::configure(basePath: dirname(__DIR__))
    ->withRouting(
        commands: __DIR__.'/../routes/console.php',
        health: '/up',
        provider: \App\Providers\RouteServiceProvider::class
    )
``` 
will totally understand this provider will be instead of the framework's one

**Comment by rodrigopedra** [comment]
> Based on the current locale of the request, they load the correct cached routes file. 

This is working as intended, right?

> So, we have overridden the `RouteServiceProvider`.

Can you provide your overridden method, so we can better understand what you are trying to accomplish?

**Comment by ahmedabdel3al** [comment]
@rodrigopedra 

> > Based on the current locale of the request, they load the correct cached routes file.
> 
> This is working as intended, right?  Yes
> 
> > So, we have overridden the `RouteServiceProvider`.
> 
> Can you provide your overridden method, so we can better understand what you are trying to accomplish? 

https://github.com/mcamara/laravel-localization/blob/master/src/Mcamara/LaravelLocalization/Traits/LoadsTranslatedCachedRoutes.php


```
/**
     * Load the cached routes for the application.
     *
     * @return void
     */
    protected function loadCachedRoutes()
    {
        $localization = $this->getLaravelLocalization();

        // compute $locale from url.
        // It is null if url does not contain locale.
        $locale = $localization->getInversedLocaleFromMapping(
            $localization->setLocale()
        );

        $localeKeys = $localization->getSupportedLanguagesKeys();

        // First, try to load the routes specifically cached for this locale
        // if they do not exist, write a warning to the log and load the default
        // routes instead. Note that this is guaranteed to exist, because the
        // 'cached routes' check in the Application checks its existence.

        $path = $this->makeLocaleRoutesPath($locale, $localeKeys);

        if ( ! file_exists($path)) {

            Log::warning("Routes cached, but no cached routes found for locale '{$locale}'!");

            $path = $this->getDefaultCachedRoutePath();
        }

        $this->app->booted(function () use ($path) {
            require $path;
        });
    }

    /**
     * Returns the path to the cached routes file for a given locale.
     *
     * @param string   $locale
     * @param string[] $localeKeys
     * @return string
     */
    protected function makeLocaleRoutesPath($locale, $localeKeys)
    {
        $path = $this->getDefaultCachedRoutePath();

        if ( ! $locale || ! in_array($locale, $localeKeys)) {
            return $path;
        }

        return substr($path, 0, -4) . '_' . $locale . '.php';
    }

    /**
     * Returns the path to the standard cached routes file.
     *
     * @return string
     */
    protected function getDefaultCachedRoutePath()
    {
        return $this->app->getCachedRoutesPath();
    }

    /**
     * @return string|null
     */
    protected function getLocaleFromRequest()
    {
        return request()->segment(1);
    }

    /**
     * @return \Mcamara\LaravelLocalization\LaravelLocalization
     */
    protected function getLaravelLocalization()
    {
        return app('laravellocalization');
    }
```


i prefer if you can also look in https://github.com/mcamara/laravel-localization?tab=readme-ov-file#caching-routes

**Comment by rodrigopedra** [comment]
Ok, I get it now. Sorry for the insistence.

From that repo's issue tracker, it seems people are finding better results with another package:

https://github.com/mcamara/laravel-localization/issues/899#issuecomment-2028882315

Maybe, it is worth taking a look in the meanwhile.

**Comment by ahmedabdel3al** [comment]
Even if there's another package that can perform the same function, 
the issue of overriding this `RouteServiceProvider` will remain. It was a feature that the framework gave me full permissions to override. After the latest release, this feature has been deprived.

**Comment by taylorotwell** [comment]
I think there is probably a better solution to this. Even something like loadCachedRoutesUsing or some sort of specific solution.

**Pull request closed without merging**